Writefreely import

joyeusenoelle opened the pull request:
Adapts the existing writeas import tool to work with any WriteFreely instance.

This tool will prompt for login information if the instance isn't stored in the included INI file. It does not currently support the config being developed elsewhere; I have plans to include that in a later revision.

It also detects blog-name collisions and adds a "friendly" 4-character string to the end of each blog's name if a blog with that name already exists in the database.

joyeusenoelle committed:
Restore to working state

joyeusenoelle committed:
Include basic instances.ini

joyeusenoelle committed:
Add comments to instances.ini

joyeusenoelle committed:
Create new token if token doesn't exist

joyeusenoelle committed:
Add more robust error message when rate limited

joyeusenoelle committed:
Update interactive section to use scanner

joyeusenoelle committed:
Separate empty line check from comment check

joyeusenoelle committed:
Amend collection name to reduce collision chance

joyeusenoelle committed:
Truncate language to 2 characters

joyeusenoelle committed:
Change blog name only on collision

joyeusenoelle committed:
Verify that it's a WordPress file

joyeusenoelle committed:
Add general struct for imported blogs

joyeusenoelle commented:
wp-import no longer simply trusts the user to feed it the right file, and now verifies that the file it's importing is a WordPress file, by checking the file extension and searching the first 200 bytes for the word "WordPress" (this appears in the comment block at the top of all WP exports).

I've also generalized the main function so that it takes a (new) ImportedBlogs struct (defined here) instead of a WXR struct, and written the function to convert the WXR struct to the new ImportedBlog format. This will make generalizing the tool to import other formats (zip, JSON, etc.) easier.

thebaer commented on main.go:
```diff
@@ -11,75 +11,148 @@
+//   wp-import -i instance filename
+//   wp-import -i instance -f filename
+
+func parseArgs(args []string) map[string]string {
```
I think we can get rid of this func and make parsing less error-prone by using the built-in `flag` package. Here's a good example of how that can be done: https://github.com/writeas/wf-migrate/blob/master/cmd/wfimport/main.go

joyeusenoelle replied:
After seeing your use of `flag` there I was planning on switching to it. :)

thebaer commented on main.go:
```diff
@@ -11,75 +11,148 @@
-		errQuit("usage: wp-import filename.xml")
+// Print the usage spec to the terminal and exit cleanly
+func printUsage(help bool) {
+	usage := "usage: wp-import [-h|--help] [-i instance] [-f] filename.xml"
```
As a bonus of using the `flag` package (as mentioned in another comment), we'll get all this output for free. So we probably won't need this func either.

thebaer reviewed (commented):
Thanks @joyeusenoelle, this is great progress. I haven't run this yet, but noticed a few things we should address after reading through the code.

robjloranger commented on main.go:
```diff
@@ -98,18 +171,181 @@ func main() {
+
+func main() {
+	a := parseArgs(os.Args)
+	// if len(os.Args) < 2 {
```
probably don't need the old code after moving this into `parseArgs`

robjloranger commented on main.go:
```diff
@@ -98,18 +171,181 @@ func main() {
+	}
+
+	instances := importConfig()
+	//fmt.Println(instances)
```
old comment

robjloranger commented on main.go:
```diff
@@ -98,18 +171,181 @@ func main() {
+		fmt.Print("Instance URL: ")
+		r.Scan()
+		url := r.Text()
+		if string(url[:5]) != "https" {
```
this could be messy if the url was somehow http

joyeusenoelle replied:
Better to check for `url[:4] != "http"` ?

robjloranger replied:
or maybe both I suppose, and then the question is should we assume https? some people have old set ups. usually https sites will redirect plain requests or still support http. but https only will not always support http requests.

joyeusenoelle replied:
I've modified to prompt the user to include http/https, and then assume https if they don't.

robjloranger replied:
nice, good call

robjloranger commented on main.go:
```diff
@@ -98,18 +171,181 @@ func main() {
+		if string(url[len(url)-1:]) == "/" {
+			url = string(url[:len(url)-1])
+		}
+		//fmt.Println("Using URL", url)
```
this and the next two commented `Println`s, are they meant to be printed or were they for debugging?

joyeusenoelle replied:
Debugging, yes.

joyeusenoelle committed:
Use flag package, add dry-run option

joyeusenoelle committed:
Change instance URL behavior

joyeusenoelle commented:
I've modified the code to use the `flag` package, and removed the now-extraneous `parseArgs` and `printUsage` functions.

I've also added the capacity to do a dry run, so users can verify that their file parses correctly before they upload everything to the remote server. (Also useful for testing changes that don't affect the upload process. ;)

joyeusenoelle committed:
Add verbose flag

joyeusenoelle committed:
Add initial Zip parsing (from zip-import)
Final commit of the day 6/21. Code doesn't currently work, but I see how to get it there.

joyeusenoelle reviewed (commented):
Ignore comment on 67-68 - I changed my mind. We're exporting ParseWPFile so it should handle the WXR stuff and not rely on the importing script to do that. ParseWPFile now takes `[]byte` as an argument.

joyeusenoelle committed:
Add ParseWFJSONFile
